### DISCUSSION TITLE
How should I properly package a task to use task isolation?

### DISCUSSION THREAD

**Discussion by bording** [dotnet/msbuild#9347] [Q&A] [answered]
I know changes have been made to load [tasks in isolated way](https://github.com/dotnet/msbuild/blob/main/documentation/specs/task-isolation-and-dependencies.md) by attempting to load each task into a custom `AsssemblyLoadContext`, and I can see that [MSBuildLoadContext](https://github.com/dotnet/msbuild/blob/221fd2e8790a22ead513eb71630557efc02060e2/src/Shared/MSBuildLoadContext.cs) exists to implement it.

However, what I've not been able to find so far is any sort of clear documentation or examples of what all I need to do when packaging a task assembly to make sure that I'm properly opting in to using it.

The layout I'm currently using is:

![image](https://github.com/dotnet/msbuild/assets/753669/8a467dbf-d888-40d0-bc82-d621241c7b18)

However, it's not clear to me if that is correct or even working. Looking at a binlog from a build where the task is being used, I see log entries like:
```
Assembly loaded during TaskRun: NServiceBus.Persistence.Sql.ScriptBuilderTask, Version=8.0.0.0, Culture=neutral, PublicKeyToken=9fc386479f8a226c (location: F:\.nuget\packages\nservicebus.persistence.sql\8.0.0-alpha.1.15\build\net8.0\task\net8.0\NServiceBus.Persistence.Sql.ScriptBuilderTask.dll, MVID: 6a4fa7fe-ac11-4dd8-9fc0-0279210ae827, AppDomain: [Default])
```

The log messages all say `AppDomain: [Default]`. Does that mean that my task assembly and its dependencies aren't being isolated?

**Comment by baronfel** [comment]
Can you take a look at [this example documentation](https://learn.microsoft.com/en-us/visualstudio/msbuild/tutorial-custom-task-code-generation?view=vs-2022#package-the-task-for-distribution) and see if that is the kind of content you expected to see for isolated task packaging?

**Comment by rainersigwald** [comment]
You shouldn't have to do anything at the task level to get the `AssemblyLoadContext` isolation on .NET.

> The log messages all say `AppDomain: [Default]`.

That's accurate but confusing: on .NET there's only ever one AppDomain (because `AssemblyLoadContext` replaces it).

**Comment by bording** [comment]
> Can you take a look at [this example documentation](https://learn.microsoft.com/en-us/visualstudio/msbuild/tutorial-custom-task-code-generation?view=vs-2022#package-the-task-for-distribution) and see if that is the kind of content you expected to see for isolated task packaging?

@baronfel There appears to be something off with the tutorial project. The tutorial states "The dependencies of your MSBuild task must be packaged inside the package; they cannot be expressed as normal package references."

However, after creating the tutorial project and building it, I don't actually see the Microsoft.Extensions.DependencyInjection assembly in the package:

![image](https://github.com/dotnet/msbuild/assets/753669/2339fe05-18f8-4186-9666-28e15337cbb2)

That tutorial also seems to have some things other things that aren't relevant now. For example, it claims that `<GenerateDependencyFile>true</GenerateDependencyFile>` is required in the project file, but I'm seeing a deps.json file being generated without it.

**Comment by bording** [comment]
> You shouldn't have to do anything at the task level to get the `AssemblyLoadContext` isolation on .NET.

@rainersigwald Is the deps.json file required to get the isolation, or is it enough to have the dependencies next to the task assembly in the package?


> > The log messages all say `AppDomain: [Default]`.
> 
> That's accurate but confusing: on .NET there's only ever one AppDomain (because `AssemblyLoadContext` replaces it).

Given that AppDomains aren't really a thing on .NET, I was wondering if that log message was instead reporting on which `AssemblyLoadContext` it was loaded into instead. If it's not, that seems like a nice change to make.

**Reply by rainersigwald** [reply, marked as answer]
> @rainersigwald Is the deps.json file required to get the isolation, or is it enough to have the depdencies next to the task assembly in the package?

Not required unless you have native dependencies; if it's all .NET assemblies just having 'em next to your task assembly should be sufficient.

**Reply by rainersigwald** [reply]
> If it's not, that seems like a nice change to make.

Yup! https://github.com/dotnet/msbuild/issues/9348

**Comment by bording** [comment]
> Not required unless you have native dependencies; if it's all .NET assemblies just having 'em next to your task assembly should be sufficient.

If you did have native dependencies, for example if you were using LibGit2Sharp in your task, I assume at that point you'd put the entire `runtimes` folder in there as well, and the deps.json file would let it pick the correct one.

**Reply by rainersigwald** [reply]
Correct, it should be basically like https://learn.microsoft.com/dotnet/core/tutorials/creating-app-with-plugin-support.
